### ISSUE TITLE
Allow secretKeyRef for billing account id in Projects

### ISSUE THREAD

**Issue by sleepycat** [GoogleCloudPlatform/k8s-config-connector#191]
The [sample code](https://cloud.google.com/config-connector/docs/reference/resources#project_in_folder) for creating a project includes the billing id.
```apiVersion: resourcemanager.cnrm.cloud.google.com/v1beta1
kind: Project
metadata:
  annotations:
    # Replace "${ORG_ID?}" with the numeric ID for your organization
    cnrm.cloud.google.com/organization-id: "${ORG_ID?}"
    cnrm.cloud.google.com/auto-create-network: "false"
  labels:
    label-one: "value-one"
  name: project-sample-in-org
spec:
  name: Config Connector Sample
  billingAccountRef:
    # Replace "${BILLING_ACCOUNT_ID?}" with the numeric ID for your billing account
    external: "${BILLING_ACCOUNT_ID?}"
```
Requiring sensitive information makes this code difficult to share, and means managing access to the code.
Ideally we would be able to reference a value in a kubernetes secret something like what you have for [SQLUser](https://cloud.google.com/config-connector/docs/reference/resources#schema_155):
```apiVersion: resourcemanager.cnrm.cloud.google.com/v1beta1
kind: Project
metadata:
  annotations:
    # Replace "${ORG_ID?}" with the numeric ID for your organization
    cnrm.cloud.google.com/organization-id: "${ORG_ID?}"
    cnrm.cloud.google.com/auto-create-network: "false"
  labels:
    label-one: "value-one"
  name: project-sample-in-org
spec:
  name: Config Connector Sample
  billingAccountRef:
    # Replace "${BILLING_ACCOUNT_ID?}" with the numeric ID for your billing account
    external:
      valueFrom:
        secretKeyRef:
          name: billing-secret  # Name of the Secret object to extract a value from
          key: account-id
```

**Comment by caieo** [comment]
Hi @sleepycat, thanks for making this request. Can I ask how you would be sharing/duplicating this code? If you're creating a project template for people to copy and use, we have solutions that work with [kpt](https://github.com/GoogleContainerTools/kpt) that will allow you to replace the billing account field easily. Here is a link to our [simple-project solution](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/tree/master/config-connector/solutions/projects/kpt/simple-project). 


> Ideally we would be able to reference a value in a kubernetes secret something like what you have for SQLUser[...]

The way we determine whether a field can be referenced by a Kubernetes secret is by checking if the field is a sensitive field in the API schema. At the moment, `billingAccountRef` isn't considered a sensitive field so I don't think we will support adding this feature.

**Comment by sleepycat** [comment]
The plan is putting the config up in a public repo on Github, and syncing that config with a cluster using [flux](https://fluxcd.io/). I've used Sealed Secrets before to allow sensitive data to be checked in to public repos, and was assuming I would do the same here for the billing account... and then realized I can't reference a secret there.

Given that you're saying that the billing account isn't sensitive (which I should read as "OK to post in a public repo", right?), I guess the fact that it's a variable in the [kpt examples](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/blob/f92f86d88494ac8dacdbafdca84803be4415afbc/config-connector/solutions/projects/kpt/simple-project/Kptfile#L23-L29) is just because its, well, variable.
 
Can you expand on "checking if the field is a sensitive field in the API schema"? Knowing where to look to determine if a field is sensitive or not would be helpful, but I don't see these markings in the Rest API or config-connector docs anywhere. I think my Google-fu is failing me. Where should I be looking to see that?

**Comment by AlexBulankou** [comment]
@sleepycat , to clarify on the terminology, when we say *sensitive field* or *secret*, i.e. information that can be used to authenticate or authorize, such as API key, security certificates, passwords. BILLING_ACCOUNT_ID is not a secret in this sense because you cannot use it authenticate. 
However, "not a secret" does not mean "OK to post in public repo".  It is personally identifiable information and so you want to consider whether you want it exposed in public GitHub repo and associated with all the other information from this repo. 
You are correct that the simple-project kpt example that you referenced makes it a variable because it is expected that the users will customize the package by replacing these values.
To your other question, to check what fields that we consider good candidates to specify from K8s secrets are then ones that support `valueFrom`, for example [SQL user password](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/crds/sql_v1beta1_sqluser.yaml#L87).

**Comment by sleepycat** [comment]
Exactly the heart of the issue.
Typically anything/everything that doesn't fit in the "OK to post in public repo" is handled by kubernetes secrets.

As you pointed out the set of "things you probably don't want to post in public repo" includes two sub-categories: "things that allow you to authenticate or authorize" and "not strictly speaking a secret but...".

Config-connector currently allows using kubernetes secrets only for the first category, leaving me with a bit of a question mark hanging over handling the second. 

What I'm trying to do is demonstrate a full self-serve PR based gitops/config-connector workflow for infrastructure provisioning for my department. By it's nature it minimally implies visibility across an entire department (since the point is to demonstrate a workflow, not just show how to write the code), probably a wider visibility than I'd want for my billing account id.

Your advice to "consider whether you want it exposed" is good. Just to make what to consider explicit, [based on this](https://cloud.google.com/billing/docs/concepts#relationships-between-resources) it seems that that as an organizationally scoped resource, a billing account id would only be usable piece of information to authenticated members of the same organization with billing* permissions (and roles like project owner [which include project billing permissions](https://cloud.google.com/billing/docs/how-to/billing-access#overview-of-cloud-billing-roles-in-cloud-iam))? The potential for harm there seems to be the latitude created by your organizations granting of permissions and the resulting ability to associate that billing account id with other resources within the organization.
I don't seem to be able to leverage the id to get any useful additional information:
```
$> gcloud alpha billing accounts describe 0X0X0X-0X0X0X-0X0X0X
displayName: My Billing Account
name: billingAccounts/0X0X0X-0X0X0X-0X0X0X
```
That said, even those details are probably enough for a fairly convincing social engineering attack.
Does that fit with your understanding?

**Comment by AlexBulankou** [comment]
@sleepycat , thanks for these additional details. My understanding is that you are referring to the part of GitOps flow where your templated config was already expanded into the final values and is checked in into a repo so that a a tool, such as [Config Sync](https://cloud.google.com/kubernetes-engine/docs/add-on/config-sync/overview) can apply it to the cluster. 

At this point all the non-K8s values  (project ID, billing account ID etc) are fully visible. What you are saying is that, even though these values are not secrets, you are not comfortable exposing these values associated with the project for all users of the repo. Is this correct?

**Comment by sleepycat** [comment]
Exactly. Kubernetes secrets are the one tool I have prevent things in that repo from sharing the visiblility of the repo. The current narrow definition of "secret" leaves me with category of "not-secret but also not-public" values I'm not sure how to deal with. 
Hence the feature request: please consider a more expansive use of secrets, since they end up being used used for all non-public values... and minimally allow for secretKeyRefs at least for obvious "not-secret but also not-public" stuff like billing account id.
That would be a really valuable addition to config-connector IMO.
